src/adapter/src/coord.rs

```diff
@@ -1734,16 +1810,30 @@ impl Coordinator {
 
         let time = write_frontier.clone().into_option().expect("checked above");
         let time = time.saturating_sub(lag);
-        let max_as_of = Antichain::from_elem(time);
+        let compaction_frontier = Antichain::from_elem(time);
+
+        // We must not select an `as_of` that is beyond any times that have not yet been written to
+        // downstream materialized views. If we would, we might skip times in the output of these
+        // materialized views, violating correctness. So our chosen `as_of` must be at most the
+        // meet of the `upper`s of all dependent materialized views.
+        let id_bundle = CollectionIdBundle {
+            storage_ids: dependent_matviews,
+            ..Default::default()
+        };
+        let max_as_of = self.least_valid_write(&id_bundle);
 
-        let as_of = min_as_of.join(&max_as_of);
+        let mut as_of = min_as_of.clone();
+        as_of.join_assign(&compaction_frontier);
+        as_of.meet_assign(&max_as_of);
 
         tracing::info!(
             export_ids = %dataflow.display_export_ids(),
             %cluster_id,
             min_as_of = ?min_as_of.elements(),
+            max_as_of = ?max_as_of.elements(),
             write_frontier = ?write_frontier.elements(),
             %lag,
+            compaction_frontier = ?compaction_frontier.elements(),
             "selecting index `as_of` as {:?}",
```

Comment:
This event format makes it harder to parse `as_of` which seems to be the most important value attached to the event. I suggest rewriting this as

```rust
%lag,
compaction_frontier = ?compaction_frontier.elements(),
as_of = ?as_of.elements(),
selecting index `as_of`,
```

Reply:
I did it that way because I thought it would be easier to find the important value that way, when it's not buried alongside all the other attributes. But no strong feelings from my side.